File: util/math/src/main/java/dev/sweety/math/pool/Pooled.java
package dev.sweety.math.pool;

import java.lang.annotation.*;

/**
 * Marks a class whose instances are managed by an object pool.
 * Instances must be obtained via the pool (never {@code new}) and
 * returned via {@link Release}-annotated methods when no longer needed.
 *
 * <pre>{@code
 * @Pooled(pool = PacketBufferAllocator.class)
 * public class PacketBuffer { ... }
 * }</pre>
 */
@Documented
@Retention(RetentionPolicy.CLASS)
@Target(ElementType.TYPE)
public @interface Pooled {

    /** The allocator or pool type that manages instances of the annotated class. */
    Class<?> pool() default Void.class;
}

File: util/math/src/main/kotlin/dev/sweety/math/pool/Pooled.kt
package dev.sweety.math.pool

import kotlin.reflect.KClass

/** Marks a type whose instances are managed by an [ObjectPool] or allocator.  */
@MustBeDocumented
@Retention(AnnotationRetention.BINARY)
@Target(AnnotationTarget.CLASS)
annotation class Pooled(val pool: KClass<*> = Void::class)
